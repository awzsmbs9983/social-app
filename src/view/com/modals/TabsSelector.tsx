import React, {createRef, useRef, useMemo, useState} from 'react'
import {observer} from 'mobx-react-lite'
import {
  ScrollView,
  StyleSheet,
  Text,
  TouchableOpacity,
  TouchableWithoutFeedback,
  View,
} from 'react-native'
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
  runOnJS,
} from 'react-native-reanimated'
import {IconProp} from '@fortawesome/fontawesome-svg-core'
import {FontAwesomeIcon} from '@fortawesome/react-native-fontawesome'
import Swipeable from 'react-native-gesture-handler/Swipeable'
import LinearGradient from 'react-native-linear-gradient'
import {useStores} from '../../../state'
import {s, colors, gradients} from '../../lib/styles'
import {match} from '../../routes'

const TAB_HEIGHT = 42

export const snapPoints = [500]

export const Component = observer(() => {
  const store = useStores()
  const [closingTabIndex, setClosingTabIndex] = useState<number | undefined>(
    undefined,
  )
  const closeInterp = useSharedValue<number>(0)
  const tabsRef = useRef<ScrollView>(null)
  const tabRefs = useMemo(
    () =>
      Array.from({length: store.nav.tabs.length}).map(() =>
        createRef<Animated.View>(),
      ),
    [store.nav.tabs.length],
  )

  // events
  // =

  const onPressNewTab = () => {
    store.nav.newTab('/')
    onClose()
  }
  const onPressCloneTab = () => {
    store.nav.newTab(store.nav.tab.current.url)
    onClose()
  }
  const onPressShareTab = () => {
    onClose()
    // TODO
  }
  const onPressChangeTab = (tabIndex: number) => {
    store.nav.setActiveTab(tabIndex)
    onClose()
  }
  const doCloseTab = (index: number) => store.nav.closeTab(index)
  const onCloseTab = (tabIndex: number) => {
    setClosingTabIndex(tabIndex)
    closeInterp.value = 0
    closeInterp.value = withTiming(1, {duration: 300}, () => {
      runOnJS(setClosingTabIndex)(undefined)
      runOnJS(doCloseTab)(tabIndex)
    })
  }
  const onNavigate = (url: string) => {
    store.nav.navigate(url)
    onClose()
  }
  const onClose = () => {
    store.shell.closeModal()
  }
  const onLayout = () => {
    // focus the current tab
    const targetTab = tabRefs[store.nav.tabIndex]
    if (tabsRef.current && targetTab.current) {
      targetTab.current.measureLayout?.(
        tabsRef.current,
        (_left: number, top: number) => {
          tabsRef.current?.scrollTo({y: top, animated: false})
        },
        () => {},
      )
    }
  }

  // rendering
  // =

  const FatMenuItem = ({
    icon,
    label,
    url,
    gradient,
  }: {
    icon: IconProp
    label: string
    url: string
    gradient: keyof typeof gradients
  }) => (
    <TouchableOpacity
      style={styles.fatMenuItem}
      onPress={() => onNavigate(url)}>
      <LinearGradient
        style={[styles.fatMenuItemIconWrapper]}
        colors={[gradients[gradient].start, gradients[gradient].end]}
        start={{x: 0, y: 0}}
        end={{x: 1, y: 1}}>
        <FontAwesomeIcon icon={icon} style={styles.fatMenuItemIcon} size={24} />
      </LinearGradient>
      <Text style={styles.fatMenuItemLabel}>{label}</Text>
    </TouchableOpacity>
  )

  const renderSwipeActions = () => {
    return <View style={[s.p2]} />
  }

  const currentTabIndex = store.nav.tabIndex
  const closingTabAnimStyle = useAnimatedStyle(() => ({
    height: TAB_HEIGHT * (1 - closeInterp.value),
    opacity: 1 - closeInterp.value,
    marginBottom: 4 * (1 - closeInterp.value),
  }))
  return (
    <View onLayout={onLayout}>
      <View style={[s.p10, styles.section]}>
        <View style={styles.fatMenuItems}>
          <FatMenuItem icon="house" label="Feed" url="/" gradient="primary" />
          <FatMenuItem
            icon="bell"
            label="Notifications"
            url="/notifications"
            gradient="purple"
          />
          <FatMenuItem
            icon={['far', 'user']}
            label="My Profile"
            url="/"
            gradient="blue"
          />
          <FatMenuItem icon="gear" label="Settings" url="/" gradient="blue" />
        </View>
      </View>
      <View style={[s.p10, styles.section]}>
        <View style={styles.btns}>
          <TouchableWithoutFeedback onPress={onPressShareTab}>
            <View style={[styles.btn]}>
              <View style={styles.btnIcon}>
                <FontAwesomeIcon size={16} icon="share" />
              </View>
              <Text style={styles.btnText}>Share</Text>
            </View>
          </TouchableWithoutFeedback>
          <TouchableWithoutFeedback onPress={onPressCloneTab}>
            <View style={[styles.btn]}>
              <View style={styles.btnIcon}>
                <FontAwesomeIcon size={16} icon={['far', 'clone']} />
              </View>
              <Text style={styles.btnText}>Clone tab</Text>
            </View>
          </TouchableWithoutFeedback>
          <TouchableWithoutFeedback onPress={onPressNewTab}>
            <View style={[styles.btn]}>
              <View style={styles.btnIcon}>
                <FontAwesomeIcon size={16} icon="plus" />
              </View>
              <Text style={styles.btnText}>New tab</Text>
            </View>
          </TouchableWithoutFeedback>
        </View>
      </View>
      <View style={[s.p10, styles.section, styles.sectionGrayBg]}>
        <ScrollView ref={tabsRef} style={styles.tabs}>
          {store.nav.tabs.map((tab, tabIndex) => {
            const {icon} = match(tab.current.url)
            const isActive = tabIndex === currentTabIndex
            const isClosing = closingTabIndex === tabIndex
            return (
              <Swipeable
                key={tab.id}
                renderLeftActions={renderSwipeActions}
                renderRightActions={renderSwipeActions}
                leftThreshold={100}
                rightThreshold={100}
                onSwipeableWillOpen={() => onCloseTab(tabIndex)}>
                <Animated.View
                  style={[
                    styles.tabOuter,
                    isClosing ? closingTabAnimStyle : undefined,
                  ]}>
                  <Animated.View
                    ref={tabRefs[tabIndex]}
                    style={[
                      styles.tab,
                      styles.existing,
                      isActive && styles.active,
                    ]}>
                    <TouchableWithoutFeedback
                      onPress={() => onPressChangeTab(tabIndex)}>
                      <View style={styles.tabInner}>
                        <View style={styles.tabIcon}>
                          <FontAwesomeIcon size={20} icon={icon} />
                        </View>
                        <Text
                          ellipsizeMode="tail"
                          numberOfLines={1}
                          suppressHighlighting={true}
                          style={[
                            styles.tabText,
                            isActive && styles.tabTextActive,
                          ]}>
                          {tab.current.title || tab.current.url}
                        </Text>
                      </View>
                    </TouchableWithoutFeedback>
                    <TouchableWithoutFeedback
                      onPress={() => onCloseTab(tabIndex)}>
                      <View style={styles.tabClose}>
                        <FontAwesomeIcon
                          size={14}
                          icon="x"
                          style={styles.tabCloseIcon}
                        />
                      </View>
                    </TouchableWithoutFeedback>
                  </Animated.View>
                </Animated.View>
              </Swipeable>
            )
          })}
        </ScrollView>
      </View>
    </View>
  )
})

const styles = StyleSheet.create({
  section: {
    borderBottomColor: colors.gray2,
    borderBottomWidth: 1,
  },
  sectionGrayBg: {
    backgroundColor: colors.gray1,
  },
  fatMenuItems: {
    flexDirection: 'row',
    marginTop: 10,
    marginBottom: 10,
  },
  fatMenuItem: {
    width: 90,
    alignItems: 'center',
    marginRight: 6,
  },
  fatMenuItemIconWrapper: {
    borderRadius: 6,
    width: 60,
    height: 60,
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: 5,
    shadowColor: '#000',
    shadowOpacity: 0.2,
    shadowOffset: {width: 0, height: 2},
    shadowRadius: 2,
  },
  fatMenuItemIcon: {
    color: colors.white,
  },
  fatMenuItemLabel: {
    fontSize: 13,
  },
  tabs: {
    height: 240,
  },
  tabOuter: {
    height: TAB_HEIGHT + 4,
    overflow: 'hidden',
  },
  tab: {
    flexDirection: 'row',
    height: TAB_HEIGHT,
    backgroundColor: colors.gray1,
    alignItems: 'center',
    borderRadius: 4,
  },
  tabInner: {
    flexDirection: 'row',
    flex: 1,
    alignItems: 'center',
    paddingLeft: 12,
    paddingVertical: 12,
  },
  existing: {
    borderColor: colors.gray4,
    borderWidth: 1,
  },
  active: {
    backgroundColor: colors.white,
    borderColor: colors.black,
    borderWidth: 1,
  },
  tabIcon: {},
  tabText: {
    flex: 1,
    paddingHorizontal: 10,
    fontSize: 16,
  },
  tabTextActive: {
    fontWeight: '500',
  },
  tabClose: {
    paddingVertical: 16,
    paddingRight: 16,
  },
  tabCloseIcon: {
    color: '#655',
  },
  btns: {
    flexDirection: 'row',
    paddingTop: 2,
  },
  btn: {
    flexDirection: 'row',
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: colors.gray1,
    borderRadius: 4,
    marginRight: 5,
    paddingLeft: 12,
    paddingRight: 16,
    paddingVertical: 10,
  },
  btnIcon: {
    marginRight: 8,
  },
  btnText: {
    fontWeight: '500',
    fontSize: 16,
  },
})
