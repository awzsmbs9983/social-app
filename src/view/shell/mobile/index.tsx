import React, {useState, useEffect, useRef} from 'react'
import {observer} from 'mobx-react-lite'
import {
  useWindowDimensions,
  FlatList,
  GestureResponderEvent,
  SafeAreaView,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native'
import {ScreenContainer, Screen} from 'react-native-screens'
import LinearGradient from 'react-native-linear-gradient'
import {GestureDetector, Gesture} from 'react-native-gesture-handler'
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
  runOnJS,
  interpolate,
} from 'react-native-reanimated'
import {FontAwesomeIcon} from '@fortawesome/react-native-fontawesome'
import {IconProp} from '@fortawesome/fontawesome-svg-core'
import {useStores} from '../../../state'
import {NavigationModel} from '../../../state/models/navigation'
import {match, MatchResult} from '../../routes'
import {Login} from '../../screens/Login'
import {Onboard} from '../../screens/Onboard'
import {Modal} from '../../com/modals/Modal'
import {MainMenu} from './MainMenu'
import {TabsSelector} from './TabsSelector'
import {Composer} from './Composer'
import {s, colors} from '../../lib/styles'
import {
  GridIcon,
  HomeIcon,
  MangifyingGlassIcon,
  BellIcon,
} from '../../lib/icons'

const SWIPE_GESTURE_DIST_TRIGGER = 0.4
const SWIPE_GESTURE_VEL_TRIGGER = 2500

const Btn = ({
  icon,
  inactive,
  notificationCount,
  onPress,
  onLongPress,
}: {
  icon: IconProp | 'menu' | 'house' | 'bell' | 'search'
  inactive?: boolean
  notificationCount?: number
  onPress?: (event: GestureResponderEvent) => void
  onLongPress?: (event: GestureResponderEvent) => void
}) => {
  let size = 21
  let addedStyles
  let IconEl
  if (icon === 'menu') {
    IconEl = GridIcon
  } else if (icon === 'house') {
    IconEl = HomeIcon
    size = 24
  } else if (icon === 'search') {
    IconEl = MangifyingGlassIcon
    size = 24
    addedStyles = {position: 'relative', top: -1} as ViewStyle
  } else if (icon === 'bell') {
    IconEl = BellIcon
    size = 24
    addedStyles = {position: 'relative', top: -1} as ViewStyle
  } else {
    IconEl = FontAwesomeIcon
  }

  if (inactive) {
    return (
      <View style={styles.ctrl}>
        {notificationCount ? (
          <View style={styles.ctrlCount}>
            <Text style={styles.ctrlCountLabel}>{notificationCount}</Text>
          </View>
        ) : undefined}
        <IconEl
          size={size}
          style={[styles.ctrlIcon, styles.inactive, addedStyles]}
          icon={icon}
        />
      </View>
    )
  }
  return (
    <TouchableOpacity
      style={styles.ctrl}
      onPress={onLongPress ? onPress : undefined}
      onPressIn={onLongPress ? undefined : onPress}
      onLongPress={onLongPress}>
      {notificationCount ? (
        <View style={styles.ctrlCount}>
          <Text style={styles.ctrlCountLabel}>{notificationCount}</Text>
        </View>
      ) : undefined}
      <IconEl size={size} style={[styles.ctrlIcon, addedStyles]} icon={icon} />
    </TouchableOpacity>
  )
}

export const MobileShell: React.FC = observer(() => {
  const store = useStores()
  const [isMainMenuActive, setMainMenuActive] = useState(false)
  const [isTabsSelectorActive, setTabsSelectorActive] = useState(false)
  const scrollElRef = useRef<FlatList | undefined>()
  const winDim = useWindowDimensions()
  const swipeGestureInterp = useSharedValue<number>(0)
  const screenRenderDesc = constructScreenRenderDesc(store.nav)

  const onPressHome = () => {
    if (store.nav.tab.current.url === '/') {
      scrollElRef.current?.scrollToOffset({offset: 0})
    } else {
      store.nav.navigate('/')
    }
  }
  const onPressSearch = () => store.nav.navigate('/search')
  const onPressMenu = () => setMainMenuActive(true)
  const onPressNotifications = () => store.nav.navigate('/notifications')
  const onPressTabs = () => setTabsSelectorActive(true)

  const goBack = () => store.nav.tab.goBack()
  const swipeGesture = Gesture.Pan()
    .onUpdate(e => {
      if (store.nav.tab.canGoBack) {
        swipeGestureInterp.value = Math.max(e.translationX / winDim.width, 0)
      }
    })
    .onEnd(e => {
      if (
        swipeGestureInterp.value >= SWIPE_GESTURE_DIST_TRIGGER ||
        e.velocityX > SWIPE_GESTURE_VEL_TRIGGER
      ) {
        swipeGestureInterp.value = withTiming(1, {duration: 100}, () => {
          runOnJS(goBack)()
        })
      } else {
        swipeGestureInterp.value = withTiming(0, {duration: 100})
      }
    })
  useEffect(() => {
    // reset the swipe interopolation when the page changes
    swipeGestureInterp.value = 0
  }, [swipeGestureInterp, store.nav.tab.current])

  const swipeTransform = useAnimatedStyle(() => ({
    transform: [{translateX: swipeGestureInterp.value * winDim.width}],
  }))
  const swipeOpacity = useAnimatedStyle(() => ({
    opacity: interpolate(swipeGestureInterp.value, [0, 1.0], [0.6, 0.0]),
  }))

  if (!store.session.isAuthed) {
    return (
      <LinearGradient
        colors={['#007CFF', '#00BCFF']}
        start={{x: 0, y: 0.8}}
        end={{x: 0, y: 1}}
        style={styles.outerContainer}>
        <SafeAreaView style={styles.innerContainer}>
          <Login />
        </SafeAreaView>
        <Modal />
      </LinearGradient>
    )
  }
  if (store.onboard.isOnboarding) {
    return (
      <View style={styles.outerContainer}>
        <View style={styles.innerContainer}>
          <Onboard />
        </View>
      </View>
    )
  }

  return (
    <View style={styles.outerContainer}>
      <SafeAreaView style={styles.innerContainer}>
        <GestureDetector gesture={swipeGesture}>
          <ScreenContainer style={styles.screenContainer}>
            {screenRenderDesc.screens.map(
              ({Com, params, key, current, previous}) => {
                return (
                  <Screen
                    key={key}
                    style={[StyleSheet.absoluteFill]}
                    activityState={current ? 2 : previous ? 1 : 0}>
                    <Animated.View
                      style={
                        current ? [styles.screenMask, swipeOpacity] : undefined
                      }
                    />
                    <Animated.View
                      style={[
                        s.flex1,
                        styles.screen,
                        current ? swipeTransform : undefined,
                      ]}>
                      <Com
                        params={params}
                        visible={true}
                        scrollElRef={current ? scrollElRef : undefined}
                      />
                    </Animated.View>
                  </Screen>
                )
              },
            )}
          </ScreenContainer>
        </GestureDetector>
      </SafeAreaView>
      <View style={styles.bottomBar}>
        <Btn icon="house" onPress={onPressHome} />
        <Btn icon="search" onPress={onPressSearch} />
        <Btn icon="menu" onPress={onPressMenu} />
        <Btn
          icon="bell"
          onPress={onPressNotifications}
          notificationCount={store.me.notificationCount}
        />
        <Btn icon={['far', 'clone']} onPress={onPressTabs} />
      </View>
      <MainMenu
        active={isMainMenuActive}
        onClose={() => setMainMenuActive(false)}
      />
      <Modal />
      <TabsSelector
        active={isTabsSelectorActive}
        onClose={() => setTabsSelectorActive(false)}
      />
      <Composer
        active={store.shell.isComposerActive}
        onClose={() => store.shell.closeComposer()}
        winHeight={winDim.height}
        replyTo={store.shell.composerOpts?.replyTo}
        onPost={store.shell.composerOpts?.onPost}
      />
    </View>
  )
})

/**
 * This method produces the information needed by the shell to
 * render the current screens with screen-caching behaviors.
 */
type ScreenRenderDesc = MatchResult & {
  key: string
  current: boolean
  previous: boolean
}
function constructScreenRenderDesc(nav: NavigationModel): {
  icon: IconProp
  screens: ScreenRenderDesc[]
} {
  let icon: IconProp = 'magnifying-glass'
  let screens: ScreenRenderDesc[] = []
  for (const tab of nav.tabs) {
    const tabScreens = [
      ...tab.getBackList(5),
      Object.assign({}, tab.current, {index: tab.index}),
    ]
    const parsedTabScreens = tabScreens.map(screen => {
      const isCurrent = nav.isCurrentScreen(tab.id, screen.index)
      const isPrevious = nav.isCurrentScreen(tab.id, screen.index + 1)
      const matchRes = match(screen.url)
      if (isCurrent) {
        icon = matchRes.icon
      }
      return Object.assign(matchRes, {
        key: `t${tab.id}-s${screen.index}`,
        current: isCurrent,
        previous: isPrevious,
      }) as ScreenRenderDesc
    })
    screens = screens.concat(parsedTabScreens)
  }
  return {
    icon,
    screens,
  }
}

const styles = StyleSheet.create({
  outerContainer: {
    height: '100%',
    flex: 1,
  },
  innerContainer: {
    flex: 1,
  },
  screenContainer: {
    flex: 1,
  },
  screen: {
    backgroundColor: colors.gray1,
  },
  screenMask: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    backgroundColor: '#000',
    opacity: 0.5,
  },
  topBar: {
    flexDirection: 'row',
    backgroundColor: colors.white,
    borderBottomWidth: 1,
    borderBottomColor: colors.gray2,
    paddingLeft: 10,
    paddingRight: 10,
    paddingTop: 40,
    paddingBottom: 5,
  },
  avi: {
    width: 34,
    height: 34,
    marginRight: 8,
    borderRadius: 17,
  },
  location: {
    flex: 1,
    flexDirection: 'row',
    borderRadius: 6,
    paddingLeft: 12,
    paddingRight: 6,
    paddingTop: 9,
    paddingBottom: 9,
    backgroundColor: colors.gray1,
  },
  locationIcon: {
    color: colors.gray5,
    marginTop: 3,
    marginRight: 6,
  },
  locationIconNudgeUp: {
    marginTop: 2,
  },
  locationIconLight: {
    color: colors.gray5,
    marginTop: 2,
    marginRight: 8,
  },
  locationText: {
    color: colors.black,
  },
  locationTextLight: {
    color: colors.gray4,
  },
  topBarBtn: {
    marginLeft: 8,
    justifyContent: 'center',
    borderRadius: 6,
    paddingHorizontal: 6,
  },
  bottomBar: {
    flexDirection: 'row',
    backgroundColor: colors.white,
    borderTopWidth: 1,
    borderTopColor: colors.gray2,
    paddingLeft: 5,
    paddingRight: 15,
    paddingBottom: 20,
  },
  ctrl: {
    flex: 1,
    paddingTop: 15,
    paddingBottom: 15,
  },
  ctrlCount: {
    position: 'absolute',
    left: 46,
    top: 10,
    backgroundColor: colors.red3,
    paddingHorizontal: 4,
    paddingBottom: 1,
    borderRadius: 8,
  },
  ctrlCountLabel: {
    fontSize: 12,
    fontWeight: 'bold',
    color: colors.white,
  },
  ctrlIcon: {
    color: colors.black,
    marginLeft: 'auto',
    marginRight: 'auto',
  },
  inactive: {
    color: colors.gray3,
  },
})
