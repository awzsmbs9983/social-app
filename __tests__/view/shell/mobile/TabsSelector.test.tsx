import React from 'react'
import {Animated} from 'react-native'
import renderer from 'react-test-renderer'
import {TabsSelector} from '../../../../src/view/shell/mobile/TabsSelector'
// import {render} from '../../../../jest/test-utils'

describe('TabsSelector', () => {
  const mockedProps = {
    active: true,
    tabMenuInterp: new Animated.Value(0),
    onClose: jest.fn(),
  }
  it('renders correctly', () => {
    const tree = renderer.create(<TabsSelector {...mockedProps} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
