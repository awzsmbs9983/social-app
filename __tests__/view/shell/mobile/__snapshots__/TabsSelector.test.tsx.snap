// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabsSelector renders correctly 1`] = `
<View
  collapsable={false}
  style={
    Object {
      "backgroundColor": "#fff",
      "borderTopColor": "#e4e2e2",
      "borderTopWidth": 1,
      "bottom": 55,
      "height": 320,
      "opacity": 1,
      "position": "absolute",
      "transform": Array [
        Object {
          "translateY": 320,
        },
      ],
      "width": "100%",
    }
  }
>
  <View
    onLayout={[Function]}
  >
    <View
      style={
        Array [
          Object {
            "padding": 10,
          },
          Object {
            "borderBottomColor": "#e4e2e2",
            "borderBottomWidth": 1,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "paddingTop": 2,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#f8f3f3",
                "borderRadius": 4,
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginRight": 5,
                "paddingLeft": 12,
                "paddingRight": 16,
                "paddingVertical": 10,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "marginRight": 8,
              }
            }
          >
            <
              icon="share"
              size={16}
            />
          </View>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                },
                Object {
                  "fontSize": 16,
                  "fontWeight": "500",
                },
              ]
            }
          >
            Share
          </Text>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#f8f3f3",
                "borderRadius": 4,
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginRight": 5,
                "paddingLeft": 12,
                "paddingRight": 16,
                "paddingVertical": 10,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "marginRight": 8,
              }
            }
          >
            <
              icon={
                Array [
                  "far",
                  "clone",
                ]
              }
              size={16}
            />
          </View>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                },
                Object {
                  "fontSize": 16,
                  "fontWeight": "500",
                },
              ]
            }
          >
            Clone tab
          </Text>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#f8f3f3",
                "borderRadius": 4,
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginRight": 5,
                "paddingLeft": 12,
                "paddingRight": 16,
                "paddingVertical": 10,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "marginRight": 8,
              }
            }
          >
            <
              icon="plus"
              size={16}
            />
          </View>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                },
                Object {
                  "fontSize": 16,
                  "fontWeight": "500",
                },
              ]
            }
          >
            New tab
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "padding": 10,
          },
          Object {
            "borderBottomColor": "#e4e2e2",
            "borderBottomWidth": 1,
          },
          Object {
            "backgroundColor": "#f8f3f3",
          },
        ]
      }
    >
      <RCTScrollView
        style={
          Object {
            "height": 240,
          }
        }
      >
        <View>
          <View
            collapsable={false}
            forwardedRef={[Function]}
            handlerTag={1}
            handlerType="PanGestureHandler"
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onLayout={[Function]}
            style={
              Object {
                "overflow": "hidden",
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "bottom": 0,
                  "flexDirection": "row",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "transform": Array [
                    Object {
                      "translateX": -10000,
                    },
                  ],
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "padding": 2,
                    },
                  ]
                }
              />
              <View
                onLayout={[Function]}
              />
            </View>
            <View
              collapsable={false}
              style={
                Object {
                  "bottom": 0,
                  "flexDirection": "row-reverse",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "transform": Array [
                    Object {
                      "translateX": -10000,
                    },
                  ],
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "padding": 2,
                    },
                  ]
                }
              />
              <View
                onLayout={[Function]}
              />
            </View>
            <View
              collapsable={false}
              forwardedRef={[Function]}
              handlerTag={2}
              handlerType="TapGestureHandler"
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
              pointerEvents="auto"
              style={
                Object {
                  "transform": Array [
                    Object {
                      "translateX": -0,
                    },
                  ],
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 46,
                    "overflow": "hidden",
                  }
                }
              >
                <View
                  collapsable={false}
                  forwardedRef={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#ffffff",
                      "borderColor": "#000000",
                      "borderRadius": 4,
                      "borderWidth": 1,
                      "flexDirection": "row",
                      "height": 42,
                    }
                  }
                >
                  <View
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "alignItems": "center",
                        "flex": 1,
                        "flexDirection": "row",
                        "paddingLeft": 12,
                        "paddingVertical": 12,
                      }
                    }
                  >
                    <View
                      style={Object {}}
                    >
                      <
                        icon="house"
                        size={20}
                      />
                    </View>
                    <Text
                      ellipsizeMode="tail"
                      numberOfLines={1}
                      style={
                        Array [
                          Object {
                            "color": "#000000",
                          },
                          Array [
                            Object {
                              "flex": 1,
                              "fontSize": 16,
                              "paddingHorizontal": 10,
                            },
                            Object {
                              "fontWeight": "500",
                            },
                          ],
                        ]
                      }
                      suppressHighlighting={true}
                    >
                      /
                    </Text>
                  </View>
                  <View
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "paddingRight": 16,
                        "paddingVertical": 16,
                      }
                    }
                  >
                    <
                      icon="x"
                      size={14}
                      style={
                        Object {
                          "color": "#655",
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            collapsable={false}
            forwardedRef={[Function]}
            handlerTag={3}
            handlerType="PanGestureHandler"
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onLayout={[Function]}
            style={
              Object {
                "overflow": "hidden",
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "bottom": 0,
                  "flexDirection": "row",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "transform": Array [
                    Object {
                      "translateX": -10000,
                    },
                  ],
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "padding": 2,
                    },
                  ]
                }
              />
              <View
                onLayout={[Function]}
              />
            </View>
            <View
              collapsable={false}
              style={
                Object {
                  "bottom": 0,
                  "flexDirection": "row-reverse",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "transform": Array [
                    Object {
                      "translateX": -10000,
                    },
                  ],
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "padding": 2,
                    },
                  ]
                }
              />
              <View
                onLayout={[Function]}
              />
            </View>
            <View
              collapsable={false}
              forwardedRef={[Function]}
              handlerTag={4}
              handlerType="TapGestureHandler"
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
              pointerEvents="auto"
              style={
                Object {
                  "transform": Array [
                    Object {
                      "translateX": -0,
                    },
                  ],
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 46,
                    "overflow": "hidden",
                  }
                }
              >
                <View
                  collapsable={false}
                  forwardedRef={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#f8f3f3",
                      "borderColor": "#968d8d",
                      "borderRadius": 4,
                      "borderWidth": 1,
                      "flexDirection": "row",
                      "height": 42,
                    }
                  }
                >
                  <View
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "alignItems": "center",
                        "flex": 1,
                        "flexDirection": "row",
                        "paddingLeft": 12,
                        "paddingVertical": 12,
                      }
                    }
                  >
                    <View
                      style={Object {}}
                    >
                      <
                        icon="bell"
                        size={20}
                      />
                    </View>
                    <Text
                      ellipsizeMode="tail"
                      numberOfLines={1}
                      style={
                        Array [
                          Object {
                            "color": "#000000",
                          },
                          Array [
                            Object {
                              "flex": 1,
                              "fontSize": 16,
                              "paddingHorizontal": 10,
                            },
                            false,
                          ],
                        ]
                      }
                      suppressHighlighting={true}
                    >
                      /notifications
                    </Text>
                  </View>
                  <View
                    accessible={true}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "paddingRight": 16,
                        "paddingVertical": 16,
                      }
                    }
                  >
                    <
                      icon="x"
                      size={14}
                      style={
                        Object {
                          "color": "#655",
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</View>
`;
